cupy/_core/internal.pyx

```diff
@@ -331,20 +328,24 @@ cdef _broadcast_core(list arrays, shape_t& shape):
 
     shape.reserve(nd)
     for i in range(nd):
-        smin = PY_SSIZE_T_MAX
-        smax = 0
+        s = 1
         for j in index:
             a = arrays[j]
             a_ndim = <Py_ssize_t>a._shape.size()
-            if i >= nd - a_ndim:
-                s = a._shape[i - (nd - a_ndim)]
-                smin = min(smin, s)
-                smax = max(smax, s)
-        if smin == 0 and smax > 1:
+            if i < nd - a_ndim:
+                continue
+            a_sh = a._shape[i - (nd - a_ndim)]
+            if a_sh == s or a_sh == 1:
+                continue
+            if s == 1:
+                s = a_sh
+                continue
             raise ValueError(
-                'shape mismatch: objects cannot be broadcast to a '
-                'single shape')
```

Comment:
FYI, the error message (added in #543) seems from [`PyArray_Broadcast`](https://numpy.org/doc/stable/reference/c-api/array.html#c.PyArray_Broadcast) in NumPy C-API.
https://github.com/numpy/numpy/blob/v1.21.1/numpy/core/src/multiarray/iterators.c#L1159-L1162